rllib/policy/sample_batch.py

```diff
@@ -133,23 +133,43 @@ def __init__(self, *args, **kwargs):
         if self._is_training is None:
             self._is_training = self.pop("is_training", False)
 
+        # Try to infer the "length" of the SampleBatch by finding the first
+        # value that is actually a ndarray/tensor.
         lengths = []
         copy_ = {k: v for k, v in self.items() if k != SampleBatch.SEQ_LENS}
         for k, v in copy_.items():
             assert isinstance(k, str), self
 
+            # If this is a nested dict (for example a nested observation),
+            # try to flatten it, assert that all elements have the same length (batch
+            # dimension)
+            if isinstance(v, (dict, tuple)):
+                if k == SampleBatch.INFOS:
+                    # Don't attempt to count on infos since we make no assumptions
+                    # about its content
+                    continue
+                v = tree.flatten(v)
+                try:
+                    # Add one of the elements length, since they are all the same
+                    len_ = len(v[0])
+                    if len_:
+                        lengths.append(len_)
+                except Exception:
+                    pass
+                else:
+                    # If we were able to figure out a length, all lengths of
+                    # elements of nested structure must be the same
+                    assert all(len(sub_space) == len(v[0]) for sub_space in v)
+                continue
+
             # TODO: Drop support for lists as values.
             # Convert lists of int|float into numpy arrays make sure all data
             # has same length.
             if isinstance(v, list):
                 self[k] = np.array(v)
 
-            # Try to infer the "length" of the SampleBatch by finding the first
-            # value that is actually a ndarray/tensor. This would fail if
-            # all values are nested dicts/tuples of more complex underlying
-            # structures.
             try:
```

Comment:
You have redundancy of code here. Can you re-write this in a way you don't repeat yourself?

Reply:
I don't think that's possible because of the `else:` after the try/except.

Reply:
How about this:
```
if k == SampleBatch.INFOS: continue

if isinstance(v, list):
    self[k] = np.array(v)
else:
    v_flat = tree.flatten(v) if isinstance(v, (dict, tuple)) else [v]
    try:
        len_ = len(v_flat[0])
        if len_:
            lengths.append(len_)
    except Exception:
        pass
    else:
        assert all(len(sub_space) == len(v_flat[0]) for sub_space in v_flat)
```

Reply:
I think the else
```
if isinstance(v, list):
    self[k] = np.array(v)
else: <<<---
```
does not make sense in this case but I get the idea, thanks!

Reply:
Also PyCharm just reminded me that `if k == SampleBatch.INFOS: continue` is not PEP conform

Reply:
I have one concern here: What if we have a sample batch that carries RNN states?
In those batches, the e.g. OBS usually have structure `[BxT, h]` (h = obs dim), whereas the "state_in_0" key has a structure of only `[B, h]` (or even more complicated: `[BxT/[max-seq-len], h]`). If this new counting procedure now finds one of the state fields first and use the len of the data for its own "count", then we get the count wrong.
Rough fix would be to skip `state_(in|out)_\d+` keys, just like we do with INFOS.

Reply:
@sven1977 That would certainly make it more clean.
On master, we always use the first found length.
I'm guessing that this only worked because typically people put observations first.

Reply:
<img width="692" alt="Screenshot 2022-11-09 at 11 05 27" src="https://user-images.githubusercontent.com/9356806/200919234-36f4b92b-f306-4409-98e9-28f81113d2cd.png">

I'm using startswith here